examples/oscp.cpp

```diff
@@ -0,0 +1,245 @@
+//                        MFEM Example 30+ - Parallel Version
+//
+// Compile with: make oscp
+//
+// Sample runs:  mpirun -np 4 oscp -m ../data/square-disc.mesh -o 1
+//               mpirun -np 4 oscp -m ../data/square-disc.mesh -o 2
+//               mpirun -np 4 oscp -m ../data/square-disc.mesh -o 2 -me 1e3
+//               mpirun -np 4 oscp -m ../data/square-disc-nurbs.mesh -o 2
+//               mpirun -np 4 oscp -m ../data/star.mesh -o 2 -eo 4
+//               mpirun -np 4 oscp -m ../data/fichera.mesh -o 2 -me 1e4
+//               mpirun -np 4 oscp -m ../data/disc-nurbs.mesh -o 2
+//               mpirun -np 4 oscp -m ../data/ball-nurbs.mesh -o 2 -eo 3 -e 1e-2
+//               mpirun -np 4 oscp -m ../data/star-surf.mesh -o 2
+//               mpirun -np 4 oscp -m ../data/square-disc-surf.mesh -o 2
+//               mpirun -np 4 oscp -m ../data/amr-quad.mesh -l 2
+//
+// Description:  This is an example of adaptive mesh refinement preprocessing
+//               which lowers the data oscillation [1] to a user-defined
+//               relative threshold. There is no PDE being solved.
+//
+//               MFEM's capability to work with both conforming and
+//               nonconforming meshes is demonstrated in example 6. In some
+//               problems, the material data or loading data is not sufficiently
+//               resolved on the initial mesh. This missing fine scale data
+//               reduces the accuracy of the solution as well as the accuracy
+//               of some local error estimators. By preprocessing the mesh
+//               before the solving the PDE, many issues can be avoided.
+//
+//               [1] Morin, P., Nochetto, R. H., & Siebert, K. G. (2000).
+//                   Data oscillation and convergence of adaptive FEM. SIAM
+//                   Journal on Numerical Analysis, 38(2), 466-488.
+//
+//               [2] Mitchell, W. F. (2013). A collection of 2D elliptic
+//                   problems for testing adaptive grid refinement algorithms.
+//                   Applied mathematics and computation, 220, 350-364.
+
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+
+// Piecewise-affine function which is sometimes mesh-conforming
+double affine_function(const Vector &p)
+{
+   double x = p(0), y = p(1);
+   if (x < 0.0)
+   {
+      return 1.0 + x + y;
+   }
+   else
+   {
+      return 1.0;
+   }
+}
+
+// Piecewise-constant function which is never mesh-conforming
+double jump_function(const Vector &p)
+{
+   if (p.Normlp(2.0) > 0.4 && p.Normlp(2.0) < 0.6) { return 1; }
+   if (p.Normlp(2.0) < 0.4 || p.Normlp(2.0) > 0.6) { return 5; }
+   return 0;
```

Comment:
Is zero really the intended third value of the function (at radius == 0.4 and 0.6)? Shouldn't the second line be just "return 5"?